pkg/controllers/management/auth/globalroles/globalrole_handler_test.go

```diff
@@ -82,18 +138,367 @@ var (
+				}
+				state.crClientMock.UpdateFunc = func(cr *normanv1.ClusterRole) (*normanv1.ClusterRole, error) {
+					state.stateChanges.createdClusterRoles[cr.Name] = cr
+					return nil, nil
+				}
+			},
+			stateAssertions: func(gtsc grTestStateChanges) {
+				require.Len(gtsc.t, gtsc.createdClusterRoles, 1)
+				cr, ok := gtsc.createdClusterRoles["clusterRole"]
+				require.True(gtsc.t, ok)
+				require.Equal(gtsc.t, &readPodCR, cr)
+			},
+			globalRole: defaultGR.DeepCopy(),
+			wantError:  false,
+			condition: reducedCondition{
+				reason: ClusterRoleExists,
+				status: metav1.ConditionTrue,
+			},
+		},
+		{
+			name: "update clusterRole fails",
+			stateSetup: func(state grTestState) {
+				state.crListerMock.GetFunc = func(_, _ string) (*normanv1.ClusterRole, error) {
+					return readConfigCR.DeepCopy(), nil
+				}
+				state.crClientMock.UpdateFunc = func(cr *normanv1.ClusterRole) (*normanv1.ClusterRole, error) {
+					return nil, fmt.Errorf("error")
+				}
+			},
+			globalRole: defaultGR.DeepCopy(),
+			wantError:  true,
+			condition: reducedCondition{
+				reason: FailedToUpdateClusterRole,
+				status: metav1.ConditionFalse,
+			},
+		},
+		{
+			name: "create clusterRole fails",
+			stateSetup: func(state grTestState) {
+				state.crListerMock.GetFunc = func(_, _ string) (*normanv1.ClusterRole, error) {
+					return nil, nil
+				}
+				state.crClientMock.CreateFunc = func(cr *normanv1.ClusterRole) (*normanv1.ClusterRole, error) {
+					return nil, fmt.Errorf("error")
+				}
+			},
+			globalRole: defaultGR.DeepCopy(),
+			wantError:  true,
+			condition: reducedCondition{
+				reason: FailedToCreateClusterRole,
+				status: metav1.ConditionFalse,
+			},
+		},
+		{
+			name: "clusterRole is created",
+			stateSetup: func(state grTestState) {
+				state.crListerMock.GetFunc = func(_, _ string) (*normanv1.ClusterRole, error) {
+					return nil, nil
+				}
+				state.crClientMock.CreateFunc = func(cr *normanv1.ClusterRole) (*normanv1.ClusterRole, error) {
+					state.stateChanges.createdClusterRoles[cr.Name] = cr
+					return nil, nil
+				}
+			},
+			stateAssertions: func(gtsc grTestStateChanges) {
+				require.Len(gtsc.t, gtsc.createdClusterRoles, 1)
+				cr, ok := gtsc.createdClusterRoles["cattle-globalrole-"]
+				require.True(gtsc.t, ok)
+				require.Equal(gtsc.t, readPodCR.Rules, cr.Rules)
+			},
+			globalRole: defaultGR.DeepCopy(),
+			wantError:  false,
+			condition: reducedCondition{
+				reason: ClusterRoleExists,
+				status: metav1.ConditionTrue,
+			},
+			annotation: "cattle-globalrole-",
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			grLifecycle := globalRoleLifecycle{}
+
+			state := setupTest(t)
+			if test.stateSetup != nil {
+				test.stateSetup(state)
+			}
+
+			grLifecycle.crLister = state.crListerMock
+			grLifecycle.crClient = state.crClientMock
+			err := grLifecycle.reconcileGlobalRole(test.globalRole)
```

Comment:
Maybe highlight this line as the function under test, with anything before setup, and anything after result validation.

Just a bit of separation through empty lines might be enough for that.

Reply:
It definitely looks nicer with the separation. I have done it for all test functions in this file. Let me know what you think.

Reply:
:+1: looks good to me. thanks.